### PR TITLE
Timelock v2

### PR DESCRIPTION
**Pull request by zhoujia6139**
## Security Checklist

- [ ] 1. Re-Entrancy
- [ ] 2. Arithmetic Over/Under Flows
- [ ] 3. Unexpected Ether
- [ ] 4. Delegatecall
- [ ] 5. Default Visibilities
- [ ] 6. Entropy Illusion
- [ ] 7. External Contract Referencing
- [ ] 8. Short Address/Parameter Attack (off chain)
- [ ] 9. Unchecked CALL Return Values
- [ ] 10. Race Conditions / Front Running
- [ ] 11. Denial Of Service (DOS)
- [ ] 12. Block Timestamp Manipulation
- [ ] 13. Constructors with Care
- [ ] 14. Uninitialized Storage Pointers
- [ ] 15. Floating Points and Precision
- [ ] 16. Tx.Origin Authentication
- [ ] 17. Address.isContract Re-Entrancy via Constructor

### ⚠️ NOTES ⚠️

Make sure to think about each of these exploits in this PR.

### CODE REVIEW

File: contracts/misc/DefaultTimeLockStrategy.sol [hunk 1]
```diff
@@ -85,7 +88,7 @@ contract DefaultTimeLockStrategy is ITimeLockStrategy {
 
         timeLockParams.releaseTime +=
             currentTimestamp +
-            _updatePeriodLimit(currentTimestamp, uint128(params.amount));
+            _updatePeriodLimit(currentTimestamp, params.amount.toUint128());
```
Comment (i5possible) [inline contracts/misc/DefaultTimeLockStrategy.sol:91]: What's the difference?

File: contracts/misc/DefaultTimeLockStrategy.sol [hunk 2]
```diff
@@ -68,9 +71,9 @@ contract DefaultTimeLockStrategy is ITimeLockStrategy {
             emit PeriodReset();
         }
 
-        uint256 newTotalAmountInCurrentPeriod = totalAmountInCurrentPeriod +
+        uint128 newTotalAmountInCurrentPeriod = totalAmountInCurrentPeriod +
```
Comment (i5possible) [inline contracts/misc/DefaultTimeLockStrategy.sol:74]: Why do we change it from uint256 to uint128?

File: contracts/misc/TimeLock.sol [hunk 3]
```diff
@@ -179,6 +185,50 @@ contract TimeLock is ITimeLock, ReentrancyGuardUpgradeable, IERC721Receiver {
...
+
+        IWETH(weth).withdraw(totalAmount);
+        Helpers.safeTransferETH(msg.sender, totalAmount);
+    }
+
+    function claimPunk(uint256[] calldata agreementIds) external nonReentrant {
+        require(!frozen, "TimeLock is frozen");
+
+        IWrappedPunks WPunk = IWrappedPunks(wpunk);
+        IPunks Punk = IPunks(WPunk.punkContract());
```
Comment (i5possible) [inline contracts/misc/TimeLock.sol:214]: Would we be able to getpunk contract as WETH does?

File: contracts/protocol/libraries/logic/SupplyLogic.sol [hunk 4]
```diff
@@ -429,13 +428,26 @@ library SupplyLogic {
...
+                );
+            }
+        } else {
+            amount = params.tokenIds.length;
+        }
         DataTypes.TimeLockParams memory timeLockParams = GenericLogic
             .calculateTimeLockParams(
                 reserve,
                 DataTypes.TimeLockFactorParams({
                     assetType: DataTypes.AssetType.ERC721,
```
Comment (i5possible) [inline contracts/protocol/libraries/logic/SupplyLogic.sol:448]: Are we using ERC721 assetType when burning UniswapV3 LP token?
For other tokens, it's the count, and it's dollor value for UniswapV3 ?

Comment (zhoujia6139) [reply contracts/protocol/libraries/logic/SupplyLogic.sol:448]: Actually it's ETH value.

File: contracts/misc/TimeLock.sol [hunk 5]
```diff
@@ -179,6 +185,50 @@ contract TimeLock is ITimeLock, ReentrancyGuardUpgradeable, IERC721Receiver {
         }
     }
 
+    function claimETH(uint256[] calldata agreementIds) external nonReentrant {
+        require(!frozen, "TimeLock is frozen");
+
+        IPoolAddressesProvider addressProvider = ACL_MANAGER
+            .ADDRESSES_PROVIDER();
+        address weth = addressProvider.getWETH();
```
Comment (GopherJ) [inline contracts/misc/TimeLock.sol:193]: can we make it an immutable to save gas?

**Pull request closed without merging**